File: lua/beer-song/beer-song.lua
local beer = {}

local function bottles(count)
  if count == 0 then return 'No more bottles' end
  if count == 1 then return '1 bottle' end
  return string.format('%d bottles', count)
end

local function beers(count)
  return bottles(count) .. ' of beer'
end

local function take(count)
  if count == 0 then return 'Go to the store and buy some more' end
  if count == 1 then return 'Take it down and pass it around' end
  return 'Take one down and pass it around'
end

function beer.verse(number)
  local b = beers(number)

  return b .. ' on the wall, ' .. b:lower() .. '.\n'
           .. take(number) .. ', ' .. beers((number-1)%100):lower() .. ' on the wall.\n'
end

function beer.sing(first, last)
  local song = {}

  for v = first, last or 0, -1 do
    song[#song+1] = beer.verse(v)
  end

  return table.concat(song, '\n')
end

return beer

File: ruby/beer-song/beer_song.rb
class BeerSong
  def verse(number)
    <<~VERSE
      #{bottles(number).capitalize} of beer on the wall, #{bottles(number)} of beer.
      #{take_one_from(number)}, #{bottles((number - 1) % 100)} of beer on the wall.
    VERSE
  end

  def verses(from, to)
    from.downto(to).map { |number| verse(number) }.join("\n")
  end

  private

  def bottles(count)
    case count
    when 0
      'no more bottles'
    when 1
      '1 bottle'
    else
      "#{count} bottles"
    end
  end

  def take_one_from(number)
    if number.zero?
      'Go to the store and buy some more'
    else
      "Take #{number == 1 ? 'it' : 'one'} down and pass it around"
    end
  end
end

module BookKeeping
  VERSION = 3
end
